### ISSUE TITLE
Investigate FP16 Main Loop Non-Matmul Arithmetic on H100

### ISSUE THREAD

**Issue by jon-chuang** [Dao-AILab/flash-attention#547]
After some further thought, I think for H100, it is worth investigating using more FP16 arithmetic. That's because FP32 arithmetic on H100 is _unbelievably expensive_ (16.7x), with FP32 TFLOPs _lower than RTX 4090_ (60 v.s. 82.6) in contrast with FP16 TFLOPs (120 v.s. 82.6).

I don't think _tensor core_ is the bottleneck for H100 perf. 

While it might be that memory bandwidth is bottleneck, I believe that investigating using FP16 arithmetic - or otherwise further reducing non-matmul arithmetic - in the main loop will be extremely productive.

![image](https://github.com/Dao-AILab/flash-attention/assets/9093549/5384fa00-fbf9-47a5-9bdb-01bcbe4b5b56)

_Originally posted by @jon-chuang in https://github.com/Dao-AILab/flash-attention/issues/543#issuecomment-1721571070_

**Comment by tridao** [comment]
Yeah if inputs are fp8 then fp16 arithmetic could make sense. That's something I'm considering as I'm working on the H100 version.

**Comment by jon-chuang** [comment]
[cuDNN FP8 fused attention](https://docs.nvidia.com/deeplearning/cudnn/developer-guide/index.html#flash-fused-multi-head-att-fp8) non-matmul arithmetic is currently in FP32 as well. It might actually be the overhead when using FP8 tensor core, for which the above numbers are even more exaggerated (FP32 - 33x, FP16 - 16x more expensive).
